conjure/objectwriter.go

```diff
@@ -93,3 +130,71 @@ func writeStructMarshalInitDecls(methodBody *jen.Group, fields []*types.Field, r
 		}
 	}
 }
+
+func getFieldSafety(fieldType types.Type, safetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
+	// Check cache first
+	if cached, exists := safetyCache[fieldType]; exists {
+		return cached
+	}
+
+	// Insert placeholder to detect cycles - use UNKNOWN as safe default
+	safetyCache[fieldType] = spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+
+	// First check if the field type itself has safety annotations
+	fieldSafety := fieldType.Safety()
+	if !fieldSafety.IsUnknown() {
+		// Update cache with real result
+		safetyCache[fieldType] = fieldSafety
+		return fieldSafety
+	}
+
+	// If it's an ObjectType, recursively compute safety from its fields
+	if objectType, ok := fieldType.(*types.ObjectType); ok {
+		result := computeStructSafety(objectType.Fields, safetyCache)
+		// Update cache with computed result
+		safetyCache[fieldType] = result
+		return result
+	}
+
+	// For other types without explicit safety, return unknown so no tags are added
+	// Cache already contains UNKNOWN, so just return it
+	return spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+}
+
+func computeStructSafety(fields []*types.Field, safetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
+	// Empty struct has no information to determine safety, so it should be unknown
+	if len(fields) == 0 {
+		return spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+	}
+
+	// Default to SAFE, then find the most restrictive level
+	// Hierarchy: safe -> unannotated -> unsafe -> do-not-log
+	overallSafety := spec.New_LogSafety(spec.LogSafety_SAFE)
+
+	for _, fieldDef := range fields {
+		fieldSafety := getFieldSafety(fieldDef.Type, safetyCache)
+
+		if fieldSafety.IsUnknown() {
+			// Unannotated field "contaminates" the struct - it becomes unannotated
+			// unless we already found something more restrictive
+			if overallSafety.Value() == spec.LogSafety_SAFE {
+				overallSafety = spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+			}
+		} else {
+			// Field has explicit safety annotation
```

Comment:
```suggestion
			// Specific safety value determined for field
```

Similar to above -- "the field is annotated" indicates to me that the field itself specifies/annotates a safety value, but in this context we're actually saying that the safety for the field was computed and was a non-unknown value

Reply:
Done.